### PR TITLE
[circuit] proves an epoch key and graffiti pre-image only

### PR DESCRIPTION
**Pull request by yuriko627**
## Pull request checklist

Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [ ] Build (`yarn build`) was run locally and any changes were pushed
- [ ] Lint (`yarn lint --check`) has passed locally and any fixes were made for failures

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Issue Number: #143 

## What is the new behavior?

-
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by vivianjeng** [comment]
should also update [scripts/circuits.mjs](https://github.com/Unirep/Unirep-Social/blob/main/packages/circuits/scripts/circuits.mjs)
and the [tests](https://github.com/Unirep/Unirep-Social/tree/main/packages/circuits/test)

**Comment by yuriko627** [comment]
@vivianjeng I'm currently getting this error for the circuit.
Do you have any clue to fix this? 

`Compiling proveGraffitiPreimage_main.circom...

internal/process/esm_loader.js:74
    internalBinding('errors').triggerUncaughtException(
                              ^
{
  pos: { first_line: 3, first_column: 17, last_line: 3, last_column: 51 },
  errStr: 'Invalid Template',
  ast: {
    type: 'FUNCTIONCALL',
    name: 'ProveGraffitiPreimage',
    params: [
      {
        type: 'NUMBER',
        value: 9n,
        first_line: 3,
        first_column: 39,
        last_line: 3,
        last_column: 40
      },
      {
        type: 'NUMBER',
        value: 9n,
        first_line: 3,
        first_column: 42,
        last_line: 3,
        last_column: 43
      },
      {
        type: 'NUMBER',
        value: 32n,
        first_line: 3,
        first_column: 45,
        last_line: 3,
        last_column: 47
      },
      {
        type: 'NUMBER',
        value: 3n,
        first_line: 3,
        first_column: 49,
        last_line: 3,
        last_column: 50
      }
    ],
    first_line: 3,
    first_column: 17,
    last_line: 3,
    last_column: 51
  },
  message: 'Invalid Template',
}`

**Comment by yuriko627** [comment]
Also, I added two ts tests, `it('successfully prove graffiti preimage'..` and `it('should fail to prove a reputation with wrong graffiti preimage'..`, but let me know if there're any other tests to add.

File: packages/circuits/circuits/proveGraffitiPreimage.circom [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    3. Pre-image of graffiti
+*/
+
+include "../node_modules/circomlib/circuits/comparators.circom";
+include "../node_modules/circomlib/circuits/gates.circom";
+include "../node_modules/circomlib/circuits/mux1.circom";
+include "../node_modules/circomlib/circuits/poseidon.circom";
+include "./sparseMerkleTree.circom";
+
+template proveGraffitiPreimage(GST_tree_depth, user_state_tree_depth, epoch_tree_depth, EPOCH_KEY_NONCE_PER_EPOCH) {
```
Comment (vivianjeng) [inline packages/circuits/circuits/proveGraffitiPreimage.circom:14]: The first alphabet should be capitalized. -> `ProveGraffitiPreimage`
To match the one in `circuit.mjs`

File: packages/circuits/circuits/proveGraffitiPreimage.circom [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+    /* 1. Check if yauser exists in the Global State Tree and verify epoch key */
+    component verify_epoch_key = VerifyEpochKey(GST_tree_depth, epoch_tree_depth, EPOCH_KEY_NONCE_PER_EPOCH);
```
Comment (vivianjeng) [inline packages/circuits/circuits/proveGraffitiPreimage.circom:17]: and there is no `VerifyEpochKey` in unirep social repo.
So you can either copy-paste the lines in `ProveNegativeReputation` or `ProveSubsidyKey` or you can write another template `VerifyEpochKey` to verify epoch key.

**Comment by vivianjeng** [comment]
> @vivianjeng Also, I added two ts tests, `it('successfully prove graffiti preimage'..` and `it('should fail to prove a reputation with wrong graffiti preimage'..`, but let me know if there're any other tests to add.

It sounds good to me. 👍🏻

**Comment by yuriko627** [comment]
So now it build but the tests that I added are not passing. 
```
  2 failing

  1) Prove grafitti preimage
       successfully prove graffiti preimage:

      AssertionError: expected false to be true
      + expected - actual

      -false
      +true
      
      at Context.<anonymous> (test/proveGraffitiPreimage.test.ts:117:30)

  2) Prove grafitti preimage
       should fail to prove a reputation with wrong graffiti preimage:

      AssertionError: expected false to be true
      + expected - actual

      -false
      +true
      
      at Context.<anonymous> (test/proveGraffitiPreimage.test.ts:142:30)

File: packages/circuits/test/proveGraffitiPreimage.test.ts [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        user_tree_root: userStateRoot,
+        GST_path_index: GSTreeProof.pathIndices,
+        GST_path_elements: GSTreeProof.siblings,
+        attester_id: attesterId,
+        pos_rep: reputationRecords[attesterId].posRep,
+        neg_rep: reputationRecords[attesterId].negRep,
+        graffiti: reputationRecords[attesterId].graffiti,
+        sign_up: reputationRecords[attesterId].signUp,
+        UST_path_elements: USTPathElements,
+    }
```
Comment (vivianjeng) [inline packages/circuits/test/proveGraffitiPreimage.test.ts:64]: Sorry I missed this message
So the `graffiti_pre_image` signal is missing in the `genCircuitInput`
and the graffiti is `hash(graffitiPreImage)`
so you should set a value to graffiti preimage e.g.`graffitiPreimage = 0`
and in `reputationRecords` you should use
```ts
        const reputationRecords = {
            [attesterId]: new Reputation(
                BigInt(10),
                BigInt(0),
                hash(graffitiPreimage),
                BigInt(1)
            ),
        }
```

and also add 
```ts
graffiti_pre_image: graffitiPreimage
```
in circuit inputs

File: packages/circuits/test/proveGraffitiPreimage.test.ts [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    const epoch = 1
+    const nonce = 1
+
+    it('successfully prove graffiti preimage', async () => {
+        const attesterId = 1
+        const reputationRecords = {
+            [attesterId]: new Reputation(
+                BigInt(10),
+                BigInt(0),
+                BigInt(1),
```
Comment (vivianjeng) [inline packages/circuits/test/proveGraffitiPreimage.test.ts:102]: should be `hash(graffitiPreimage)`

File: packages/circuits/test/proveGraffitiPreimage.test.ts [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+                BigInt(1)
+            ),
+        }
+        const circuitInputs = genCircuitInput(
+            new ZkIdentity(),
+            epoch,
+            reputationRecords,
+            attesterId
+        )
```
Comment (vivianjeng) [inline packages/circuits/test/proveGraffitiPreimage.test.ts:111]: you can pass a `graffitiPreimage` as input in this function

File: packages/circuits/test/proveGraffitiPreimage.test.ts [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+        const { proof, publicSignals } = await genProofAndPublicSignals(
+            circuitInputs
+        )
+        const isValid = await verifyProof(publicSignals, proof)
+        expect(isValid).to.be.true
```
Comment (vivianjeng) [inline packages/circuits/test/proveGraffitiPreimage.test.ts:142]: it should fail the test so you should use
```ts
expect(isValid).to.be.false
```
and you should also check the `hash(preimage) !== grafiiti`

File: packages/circuits/test/proveGraffitiPreimage.test.ts [hunk 7]
```diff
@@ -61,6 +64,7 @@ const genCircuitInput = (
         graffiti: reputationRecords[attesterId].graffiti,
         sign_up: reputationRecords[attesterId].signUp,
         UST_path_elements: USTPathElements,
+        graffiti_pre_image: graffitiPreImage,
```
Comment (vivianjeng) [inline packages/circuits/test/proveGraffitiPreimage.test.ts:67]: you should use the `graffiti_pre_image` from the input instead of global `graffitiPreImage`.
so that you can generate a wrong graffiti pre-image

File: packages/circuits/test/proveGraffitiPreimage.test.ts [hunk 8]
```diff
@@ -122,24 +127,30 @@ describe('Prove grafitti preimage', function () {
...
         )
 
         const { proof, publicSignals } = await genProofAndPublicSignals(
             circuitInputs
         )
         const isValid = await verifyProof(publicSignals, proof)
-        expect(isValid).to.be.true
+        expect(isValid).to.be.false // should fail with a wrong graffiti preimage
+
+        expect(hashOne(BigInt(graffitiPreImage))).to.not.eqaul(
```
Comment (vivianjeng) [inline packages/circuits/test/proveGraffitiPreimage.test.ts:151]: `eqaul` is a typo here.
should be `equal`

File: packages/circuits/test/proveGraffitiPreimage.test.ts (hunk 7 above)
Comment (vimwitch) [reply packages/circuits/test/proveGraffitiPreimage.test.ts:66]: This needs to be addressed

File: packages/circuits/test/proveGraffitiPreimage.test.ts [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+        const reputationRecords = {
+            [attesterId]: new Reputation(
+                BigInt(10),
+                BigInt(0),
+                hashOne(BigInt(graffitiPreImage)),
+                BigInt(1)
+            ),
+        }
+        const id = new ZkIdentity()
+        const wrongGraffitiPreImage = graffitiPreImage + 1
```
Comment (vimwitch) [inline packages/circuits/test/proveGraffitiPreimage.test.ts:134]: I think this works, but you should also test with a non-zero graffiti pre-image. This is why you should pass it as a variable to `genCircuitInput` so the function can calculate the currect user/state tree leaf.

**Review by vimwitch — CHANGES_REQUESTED** [review]
I think the circuit looks correct, it just needs some more tests, mostly with non-zero graffiti cases.

File: packages/circuits/test/proveGraffitiPreimage.test.ts (hunk 9 above)
Comment (yuriko627) [reply packages/circuits/test/proveGraffitiPreimage.test.ts:163]: so should I make another argument in`genCircuitInput` for preimages and should I pass `wrongGraffitiPreimage` there for the failing test case?

Comment (yuriko627) [reply packages/circuits/test/proveGraffitiPreimage.test.ts:163]: I don't know if I interpreted comments from you guys correctly, but I made a modification about this. I also added another test with non-zero graffiti preimage. but these tests are not passing any more. any idea why??

Comment (yuriko627) [reply packages/circuits/test/proveGraffitiPreimage.test.ts:163]: @vimwitch Sorry to bother you, but I'd appreciate it if you can comment on this!

**Comment by yuriko627** [comment]
@vivianjeng now all the tests, including non-zero graffiti preimage one, pass! do they look okay now?

File: packages/circuits/test/proveGraffitiPreimage.test.ts (hunk 9 above)
Comment (vimwitch) [reply packages/circuits/test/proveGraffitiPreimage.test.ts:163]: Yep, i think it looks good. I would remove the top level `graffitiPreImage` variable and instead define it for each test function

**Review by vivianjeng — APPROVED** [review]
LGTM. Actually we also need the `core`, `backend`, `frontend` that accepts this proof.
So I will create issues for them. and I think it can be done before our v2
